### ISSUE TITLE
Remote Overrides Not Working in Dev

### ISSUE THREAD

**Issue by gestrich** [LoopKit/Loop#1536]
**Describe the bug**

Remote overrides do not work in dev. The device token is not being consistently uploaded to Nightscout leading to overrides failing.

**To Reproduce**

1. Fresh install of Loop dev branch.
2. Enter Nightscout credentials during onboarding.
3. Agree to iOS prompt to allow notifications at end of onboarding.
4. Add a few overrides to Loop and enact one of these overrides locally.
5. Go on Nightscout and attempt to enact another override. An error "Could no find deviceToken in loopSettings".

**Expected behavior**

It is expected that you can enact overrides from Nightscout after agreeing to the prompt.

**Phone**
 - Hardware: iPhone XS
 - OS Version: 14.7.1

**Loop Version**
 - Dev - latest.

**Comment by gestrich** [comment]
@ps2 Here is a write-up on what appears to be the problem. I can put up a pull requests if you like but would like your feedback beforehand.

** Dev Notes **

There are a few identified issues:

1. *LoopDataManager Currently Expects Device Token to be Persisted*:  In LoopDataManager.storeSettings(), the deviceToken is overridden by whatever is persisted in 'LoopSettings`. Since `LoopSettings` doesn't persist the `deviceToken`, that is always nil. This was discussed in a possible solution in https://github.com/LoopKit/Loop/pull/1494. That was declined though since the token should not be persisted per iOS guidelines. A simple hack may be to no longer use the LoopSettings device token like below. But is is also possible we need/want a new paradigm to pass along non-persisted values to remote services, like the deviceToken. That's more involved.

```
//deviceToken: loopSettings.deviceToken?.hexadecimalString,
deviceToken: self.settings.deviceToken?.hexadecimalString, //Not persisting the deviceToken.
```

2.  *NightScout Uploader is Created Lazily*: When onboarding, the uploader will be set to nil in its lazy getter. This leads to the upload failing when the notification prompt is agreed to.

**Comment by gestrich** [comment]
Here's how I'm handling these various issues in the remote bolus branch for now https://github.com/LoopKit/LoopWorkspace/pull/12/commits/17bb9b6d03a6e521a2580edfcee7acef667ebe0f

**Comment by avouspierre** [comment]
I tried to modify the dev version with PR 1494 + your hack but always the same issue error.  I tried to restart NS and re-compile Loop but no change. I seems unable to register the device token. Do I need to do a fresh install ?

**Comment by gestrich** [comment]
@avouspierre This commit contains the updated solution https://github.com/LoopKit/LoopWorkspace/commit/17bb9b6d03a6e521a2580edfcee7acef667ebe0f

**Comment by avouspierre** [comment]
Thank a lot @gestrich . The patch is perfect and allows to do remote overrides with NS

**Comment by blameitonben** [comment]
Looking to use this also, but not totally sure how to build with this commit, any help is appreciated. Been missing the remote overrides :)

Thanks!

Ben
________________________________

**Comment by Willbumster** [comment]
Me too!

**Comment by jonfawcett** [comment]
Is this still the only fix for remote overrides in Dev? 

Symptoms: remote overrides don't do anything after Dev install. Locally set overrides also do not upload to NS Treatments table so they are not on the NS graph. I'm not sure if both of these are caused by the same issue since I haven't found mentions of the second. I haven't started my own debugging yet to figure out what's going on because I found this old issue.

**Comment by Trpl7ca** [comment]
I don't use remote overrides, but my local overrides have not been showing up either in NS while  using dev or dash-integration.

**Comment by gestrich** [comment]
I do not think this fixes the issue of overrides not showing on the Nightscout graph (missing the tape). This fix is in the remove bolus branch. I'm not sure of other fixes or what has made it into dev.

**Comment by gestrich** [comment]
I noticed that @ps2 has fixed #2 mentioned above in dev with the below commit. I'm not sure about the other issues I mentioned in this PR.

```
commit eb97cb140a7ddc27ed497ab414c30cfba44a68d1
Author: Pete Schwamb <pete@schwamb.net>
Date:   Mon Feb 21 20:39:38 2022 -0600

    Adding fetch of therapy settings, and fixing bug with uncalled completion handlers
```

**Comment by ps2** [comment]
I believe the issues described here are now fixed. There remain reports of Loop overrides not being triggered even when the notification arrives and is displayed on the user's phone. As best I can tell, this seems to be an issue with `dasd` not allowing Loop to be launched for some reason, and this behavior appears to be resolved with iOS 15.4 (beta).

**Comment by jgefaell** [comment]
Prior to this pull and build from today I was unable to enact remote overrides from Nightscout. As of this code it is working as expected. I can enact a remote override in Nightscout and it is invoked and appears in the Loop app.

I'm not sure if this observation is part of this issue? I couldn't find another and hadn't opened one yet. 

```
* gitRevision: 7c6b7c1
* gitBranch: N/A
* workspaceGitRevision: 8d9072c
* workspaceGitBranch: dev
```
